validator/server_arb/execution_run.go

```diff
@@ -79,6 +83,96 @@ func (e *executionRun) GetStepAt(position uint64) containers.PromiseInterface[*v
 	})
 }
 
+func (e *executionRun) GetMachineHashesWithStepSize(machineStartIndex, stepSize, requiredNumHashes uint64) containers.PromiseInterface[[]common.Hash] {
+	return stopwaiter.LaunchPromiseThread(e, func(ctx context.Context) ([]common.Hash, error) {
+		return e.machineHashesWithStepSize(ctx, machineStartIndex, stepSize, requiredNumHashes)
+	})
+}
+
+func (e *executionRun) machineHashesWithStepSize(
+	ctx context.Context,
+	machineStartIndex,
+	stepSize,
+	requiredNumHashes uint64,
+) ([]common.Hash, error) {
+	if stepSize == 0 {
+		return nil, fmt.Errorf("step size cannot be 0")
+	}
+	if requiredNumHashes == 0 {
+		return nil, fmt.Errorf("required number of hashes cannot be 0")
+	}
+	machine, err := e.cache.GetMachineAt(ctx, machineStartIndex)
+	if err != nil {
+		return nil, err
+	}
+	log.Debug(fmt.Sprintf("Advanced machine to index %d, beginning hash computation", machineStartIndex))
+
+	// If the machine is starting at index 0, we always want to start at the "Machine finished" global state status
+	// to align with the machine hashes that the inbox machine will produce.
+	var machineHashes []common.Hash
+	if machineStartIndex == 0 {
+		gs := machine.GetGlobalState()
+		log.Debug(fmt.Sprintf("Start global state for machine index 0: %+v", gs))
+		machineHashes = append(machineHashes, machineFinishedHash(gs))
+	} else {
+		// Otherwise, we simply append the machine hash at the specified start index.
+		machineHashes = append(machineHashes, machine.Hash())
+	}
+	startHash := machineHashes[0]
+
+	// If we only want 1 hash, we can return early.
+	if requiredNumHashes == 1 {
+		return machineHashes, nil
+	}
+
+	logInterval := requiredNumHashes / 20 // Log every 5% progress
```

Comment:
Optional: Maybe this should be configurable? It seems like something we might want to be more noisy in the beginning, but then turn way down low later?